qualtran/bloqs/data_loading/qrom.py

```diff
@@ -70,41 +74,81 @@ class QROM(UnaryIterationGate):
             Babbush et. al. (2020). Figure 3.
     """
 
-    data: Sequence[NDArray] = attrs.field(converter=_to_tuple)
-    selection_bitsizes: Tuple[int, ...] = attrs.field(
+    data_or_shape: Union[NDArray, Shaped] = attrs.field(
+        converter=lambda x: np.array(x) if isinstance(x, (list, tuple)) else x
+    )
+    selection_bitsizes: Tuple[SymbolicInt, ...] = attrs.field(
         converter=lambda x: tuple(x.tolist() if isinstance(x, np.ndarray) else x)
     )
-    target_bitsizes: Tuple[int, ...] = attrs.field(
+    target_bitsizes: Tuple[SymbolicInt, ...] = attrs.field(
         converter=lambda x: tuple(x.tolist() if isinstance(x, np.ndarray) else x)
     )
-    num_controls: int = 0
+    num_controls: SymbolicInt = 0
+
+    def has_data(self) -> bool:
+        return not isinstance(self.data_or_shape, Shaped)
+
+    @property
+    def data_shape(self) -> Tuple[SymbolicInt, ...]:
+        return shape(self.data_or_shape)[1:]
+
+    @property
+    def data(self) -> np.ndarray:
+        if not self.has_data():
+            raise ValueError(f"Data not available for symbolic QROM {self}")
+        assert isinstance(self.data_or_shape, np.ndarray)
+        return self.data_or_shape
+
+    def __attrs_post_init__(self):
+        assert all([is_symbolic(s) or isinstance(s, int) for s in self.selection_bitsizes])
+        assert all([is_symbolic(t) or isinstance(t, int) for t in self.target_bitsizes])
+        assert len(self.target_bitsizes) == self.data_or_shape.shape[0], (
+            f"len(self.target_bitsizes)={len(self.target_bitsizes)} should be same as "
+            f"len(self.data)={self.data_or_shape.shape[0]}"
+        )
+        if isinstance(self.data_or_shape, np.ndarray) and not is_symbolic(*self.target_bitsizes):
+            assert all(
+                t >= int(np.max(d)).bit_length() for t, d in zip(self.target_bitsizes, self.data)
+            )
+        assert isinstance(self.selection_bitsizes, tuple)
+        assert isinstance(self.target_bitsizes, tuple)
 
     @classmethod
-    def build(cls, *data: ArrayLike, num_controls: int = 0) -> 'QROM':
-        _data = [np.array(d, dtype=int) for d in data]
-        selection_bitsizes = tuple((s - 1).bit_length() for s in _data[0].shape)
+    def build_from_data(cls, *data: ArrayLike, num_controls: SymbolicInt = 0) -> 'QROM':
+        _data = np.array([np.array(d, dtype=int) for d in data])
+        selection_bitsizes = tuple((s - 1).bit_length() for s in _data.shape[1:])
```

Comment:
should this raise an error if data is symbolic?

Reply:
We have the type annotation of `ArrayLike` (so mypy will complain if you pass in an object of `Shaped`) and we pass it to `np.array(d, dtype=int)` which will raise an error if `d` cannot be converted to an array of integers. So I think this is fine as is.